README.md
# Curso GRÁTIS de GNU/Linux Ubuntu Server 18.04.x LTS DevOps

Francisco Neto<br>
YouTUBE : https://www.youtube.com/c/FranciscoGPNeto<br>

Esse curso tem como objetivo de mostrar a instalação e configuração da nova versão de Servidor da Canonical o GNU/Linux Ubuntu Server 18.04.x, essa nova versão foi projetada para uma fácil implementação, administração e configuração dos serviços de rede de forma fácil e confiável.

Projetado para suportar implementações On-Premise (Servidores Locais) ou em Cloud (Nuvens), possui várias ferramentas integradas e soluções desenvolvidas pela Canonical que facilitar a implementação em larga escala e suporte a Hosts Virtual e Cloud melhorados.

Para esse curso as metodologias de DevOps, utilizando o Git e Github serão empregadas de forma dinâmica, desenvolvendo scripts e automatizando todo os processos de instalação e configuração.

Ideal para Estudantes e Profissionais de Redes que desejam se aprofundar nos pré-requisitos técnicos e práticos para Implementar Servidores GNU/Linux Ubuntu Server em cenários On-Premise ou Cloud.

## **Links Oficiais do Ubuntu, Mint, pfSense, Microsoft e VirtualBOX:**
Canonical: https://canonical.com/<br>
Linux Ubuntu: https://ubuntu.com/<br>
Linux Mint: https://www.linuxmint.com/<br>
pfSense: https://www.pfsense.org/<br>
Microsoft Windows:  https://www.microsoft.com/pt-br/windows<br>
Oracle VirtualBOX: https://www.virtualbox.org/


## **APRESENTAÇÃO do CURSO de GNU/Linux Ubuntu Server 18.04.x LTS DevOps**

DevOps Github APRESENTAÇÃO do CURSO de GNU/Linux Ubuntu Server 18.04.x LTS, vídeo mostrando os detalhes do curso e informações extras.

Link da vídeo aula (Autor: Bora Praticar): https://www.youtube.com/watch?v=qvUVWUz1MmY

## **APRESENTAÇÃO do CENÁRIO do Curso de GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando o cenário das máquinas virtuais utilizadas no Curso de DevOps Github GNU/Linux Ubuntu Server 18.04.x LTS

Link da vídeo aula: https://www.youtube.com/watch?v=8AaHNXcrxag

## **COMO instalar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o GNU/Linux Ubuntu Server 18.04.x LTS no VirtualBOX. Versões LTS (Long-term support) do Ubuntu são recomendadas devido ao suporte de 5 anos da data de lançamento. A nova versão do GNU/Linux Ubuntu Server 18.04.1 LTS trás um novo instalador, vários recursos novos e várias tecnologias que facilitam a implementação de serviços em rede, Cloud e Segurança de forma rápida, fácil e eficiente para pequenas, médias e grandes empresas.

Link da vídeo aula: https://www.youtube.com/watch?v=zDdCrqNhIXI&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Apt-Get.sh

## **COMO instalar a versão Mini (Netboot - Netinstall) do GNU/Linux Ubuntu Server 18.04 LTS**

Vídeo mostrando os procedimentos básicos para instalar a versão Mini (Netboot - Netinstall) do GNU/Linux Ubuntu Server 18.04 LTS. A versão Mini também conhecida como Netboot ou Netinstall possui os recursos mínimos para a instalação do Server ou Desktop Ubuntu, essa versão e indicada para instalação via Rede, utilizando os Protocolos DHCP, TFTP e PXE, indicada para customizações de sistema ou servidores de virtualização e cloud.

Link da vídeo aula: https://www.youtube.com/watch?v=l-AVJE3DZfw&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Netboot.sh

## **COMO atualizar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para manter o Servidor GNU/Linux Ubuntu Server 18.04.x LTS sempre atualizado.

Link da vídeo aula: https://www.youtube.com/watch?v=esnu8TAepHU&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Apt-Get.sh

## **COMO desligar e reiniciar o GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para desligar e reiniciar o Servidor GNU/Linux Ubuntu Server 18.04.x LTS

Link da vídeo aula: https://www.youtube.com/watch?v=XLFhmnKr9O0&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Shutdown.sh

## **COMO configurar a PLACA DE REDE no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar as interfaces (placa de rede) no GNU/Linux Ubuntu Server 18.04.x LTS

Na versão 18.04.x as configurações de Placa de Rede não fica mais no arquivo /etc/network/interfaces, hoje o projeto utiliza o software Netplan para fazer as configurações de rede.

Link da vídeo aula: https://www.youtube.com/watch?v=zSUd4k108Zk

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/Netplan.sh

## **COMO configurar o Hostname, Hosts e TCP Wrappers no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar os parâmetros de rede no GNU/Linux Ubuntu Server 18.04.x LTS

Link da vídeo aula: https://www.youtube.com/watch?v=J7eyb5ynjZA

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/HostnameHosts.sh

## **COMO configurar o LOCALE, TIMEZONE, UTC, UTF e KEYBOARD no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar o LOCALE (Localidade), Timezone (Fuso Horário), Data e Hora, sincronização da Data e Hora do Sistema Operacional e o Hardware (BIOS), sincronização da Data e Hora de Servidores NTP, configuração de Teclado ABNT-2, Configuração do UTF-8 no Console do Servidor GNU/Linux Ubuntu Server 18.04.x LTS

Link da vídeo aula: https://www.youtube.com/watch?v=laiuH1zgg3U&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/LocaleTimezoneKeyboard.sh 

## **SEQUÊNCIA DOS VÍDEOS DE INSTALAÇÃO DOS SERVIÇOS DE REDE NO GNU/LINUX UBUNTU SERVER 18.04.x LTS**

Recomendo seguir a sequência de vídeos abaixo para a implementação dos Serviços de Rede no Servidor Ubuntu 18.04.x, alguns script possuem dependências dos scripts anteriores, cada script complementa o serviço instalado e configurado.

## **COMO configurar o OpenSSH no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar o OpenSSH Server no GNU/Linux Ubuntu Server 18.04 LTS

OpenSSH (Open Secure Shell) é um conjunto de utilitários de rede relacionado à segurança que provém a criptografia em sessões de comunicações em uma rede de computadores usando o protocolo SSH. Foi criado com um código aberto alternativo ao código proprietário da suíte de softwares Secure Shell, oferecido pela SSH Communications Security. OpenSSH foi desenvolvido como parte do projeto OpenBSD.

Link da vídeo aula: https://www.youtube.com/watch?v=ecuol8Uf1EE

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/openssh.sh

## **COMO instalar o LAMP Server no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o LAMP Server GNU/Linux Ubuntu Server 18.04 LTS

O LAMP é o acrônimo para Linux + Apache + MySQL + PHP, é um dos principais serviços de hospedagem de páginas de internet utilizado em todo mundo, sua instalação e simples, fácil, rápida e seguro, podendo montar um estrutura de servidores de hospedagem interna ou externa.

Link da vídeo aula: https://www.youtube.com/watch?v=6EFUu-I3u4s&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/lamp.sh

## **COMO instalar o WORDPRESS no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o WORDPRESS no LAMP Server do GNU/Linux Ubuntu Server 18.04 LTS

O LAMP é o acrônimo para Linux + Apache + MySQL + PHP, é um dos principais serviços de hospedagem de páginas de internet utilizado em todo mundo, sua instalação e simples, fácil, rápida e segura, podendo montar um estrutura de servidores de hospedagem interna ou externa.

O Wordpress é um sistema livre e aberto de gestão de conteúdo para internet (do inglês: Content Management System - CMS), baseado em PHP com banco de dados MySQL, executado em um servidor interpretador, voltado principalmente para a criação de páginas eletrônicas (sites) e blogs online. Criado a partir do extinto b2/cafelog, por Ryan Boren e Matthew Mullenweg, e distribuído gratuitamente sob a GNU General Public License.

Link da vídeo aula: https://www.youtube.com/watch?v=Fs2B7kLdlm4

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/wordpress.sh

## **COMO instalar o WEBMIN no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o WEBMIN no GNU/Linux Ubuntu Server 18.04 LTS

O Webmin é um programa de gerenciamento de servidor, que roda em plataformas Unix/Linux. Com ele você pode usar também o Usermin e o Virtualmin. O Webmin funciona como um centralizador de configurações do sistema, monitoração dos serviços e de servidores, fornecendo uma interface amigável, e que quando configurado com um servidor web, pode ser acessado de qualquer local, através de um navegador: https:\\(ip do servidor):(porta de utilização). Exemplo: https:\\172.16.1.20:10000.

Link da vídeo aula: https://www.youtube.com/watch?v=rKTFI9s_YNY&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/webmin.sh

## **COMO instalar o NETDATA no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o NETDATA no GNU/Linux Ubuntu Server 18.04 LTS

O Netdata é uma ferramenta para visualizar e monitorar métricas em tempo real, otimizado para acumular todos os tipos de dados, como uso da CPU, atividade do disco, consultas SQL, visitas a um site, etc. A ferramenta é projetada para visualizar o agora com o máximo de detalhes possível, permitindo que o usuário tenha uma visão do que está acontecendo e do que acaba de acontecer em seu sistema ou aplicativo, sendo uma opção ideal para resolver problemas de desempenho em tempo real.
https:\\(ip do servidor):(porta de utilização). Exemplo: https:\\172.16.1.20:19999

Link da vídeo aula: https://www.youtube.com/watch?v=TdTTgztYVKc&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/netdata.sh

## **COMO instalar o LOGANALYZER no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o LOGANALYZER no GNU/Linux Ubuntu Server 18.04 LTS

O LogAnalyzer é uma interface da Web para o Syslog/Rsyslog e outros dados de eventos da rede. Ele fornece fácil navegação análise de eventos de rede em tempo real e serviços de relatórios. Os relatórios ajudam a manter um visão na atividade da rede. Ele consolida o Syslog/Rsyslog e outros dados de eventos, fornecendo uma página web de fácil leitura. Os gráficos  ajudam a ver as coisas importantes de relance.

Link da vídeo aula: https://www.youtube.com/watch?v=O5W25TSTJmw&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/loganalyzer.sh

## **COMO instalar o DOCKER e PORTAINER.io no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o DOCKER e o PORTAINER.io no GNU/Linux Ubuntu Server 18.04 LTS

O Docker é uma tecnologia de software que fornece contêineres, promovido pela empresa Docker, Inc. O Docker fornece uma camada adicional de abstração e automação de virtualização de nível de sistema operacional no Windows e no Linux. O Docker usa as características de isolação de recurso do núcleo do Linux como cgroups e espaços de nomes do núcleo, e um sistema de arquivos com recursos de união, como OverlayFS e outros para permitir "contêineres" independentes para executar dentro de uma única instância Linux, evitando a sobrecarga de iniciar e manter máquinas virtuais (VMs).

O Portainer.io uma solução de gerenciamento para o Docker, com ele é possível gerenciar facilmente os seus hosts Docker e clusters com Docker Swarm através de uma interface web limpa, simples e intuitiva.

Link da vídeo aula: https://www.youtube.com/watch?v=ixgLCHX_7QA

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/docker.sh

## **COMO instalar o OPENFIRE no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o OPENFIRE no GNU/Linux Ubuntu Server 18.04 LTS

Extensible Messaging and Presence Protocol (XMPP) (conhecido anteriormente como Jabber) é um protocolo aberto, extensível, baseado em XML, para sistemas de mensagens instantâneas, desenvolvido originalmente para mensagens instantâneas e informação de presença formalizado pelo IETF. Softwares com base XMPP são distribuídos em milhares de servidores através da internet, e usados por cerca de dez milhões de pessoas em todo mundo, de acordo com a  XMPP Standards Foundation.

O Openfire (anteriormente conhecido como Wildfire e Jive Messenger) é um servidor de mensagens instantâneas e de conversas em grupo que usa o servidor XMPP escrito em Java e licenciado sob a licença Apache 2.0.

Link da vídeo aula: https://www.youtube.com/watch?v=_2-cVeA-Rvo&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/openfire.sh

## **COMO instalar o ZONEMINDER no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o ZONEMINDER no GNU/Linux Ubuntu Server 18.04 LTS

ZoneMinder é um sistema de CFTV (Circuito Fechado de Televisão) Open Source, desenvolvido para sistemas operacionais Linux. Ele é liberado sob os termos da GNU General Public License (GPL). Os usuários  controlam o ZoneMinder através de uma interface baseada na Web. O aplicativo pode usar câmeras padrão (por meio de uma placa de captura, USB, Firewire etc.) ou dispositivos de câmera baseados em IP. O software permite três modos de operação: monitoramento (sem gravação), gravação após movimento detectado e gravação permanente.

CCTV / CFTV = (Closed-Circuit Television - Circuito fechado de televisão);

PTZ Pan/Tilt/Zoom (Uma câmera de rede PTZ oferece funcionalidade de vídeo em rede combinada com o recurso de movimento horizontal, vertical e de zoom - Pan = Panorâmica Horizontal - Tilt = Vertical | Zoom - Aproximar).

Link da vídeo aula: https://www.youtube.com/watch?v=YbnKB5iUCWg

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/zoneminder.sh

## **COMO instalar o ASTERISK no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o ASTERISK no GNU/Linux Ubuntu Server 18.04 LTS

O Asterisk é um software livre, de código aberto, que implementa em software os recursos encontrados em um PABX convencional, utilizando tecnologia de VoIP. Ele foi criado pelo Mark Spencer em 1999. Inicialmente desenvolvido pela empresa Digium, hoje recebe contribuições de programadores ao redor de todo o mundo. Seu desenvolvimento é ativo e sua área de aplicação muito promissora.

Link da vídeo aula: https://www.youtube.com/watch?v=IAAkmZMfby0&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/asterisk.sh

## **COMO configurar o ASTERISK no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para configurar o ASTERISK no GNU/Linux Ubuntu Server 18.04 LTS

O Asterisk é um software livre, de código aberto, que implementa em software os recursos encontrados em um PABX convencional, utilizando tecnologia de VoIP. Ele foi criado pelo Mark Spencer em 1999. Inicialmente desenvolvido pela empresa Digium, hoje recebe contribuições de programadores ao redor de todo o mundo. Seu desenvolvimento é ativo e sua área de aplicação muito promissora.

Link da vídeo aula: https://www.youtube.com/watch?v=cRCBDSOdL2E

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/AsteriskConfiguracao.sh

## **COMO instalar o OWNCLOUD no GNU/Linux Ubuntu Server 18.04.x LTS**

Vídeo mostrando os procedimentos básicos para instalar o owncloud no GNU/Linux Ubuntu Server 18.04 LTS

O ownCloud é um sistema de computador mais conhecido como "serviço de armazenamento e sincronização de arquivos". Como tal, ownCloud é muito semelhante ao amplamente usado Dropbox, cuja principal diferença é que ownCloud é gratuito e open-source, e permitindo assim qualquer um de instalar e operar sem custo em um servidor privado, sem limite de espaço de armazenamento (com exceção da capacidade do disco rígido) ou o número de clientes conectados.

Link da vídeo aula: https://www.youtube.com/watch?v=ZiOYoR4uZjU&t

Link do script utilizado nesse vídeo: https://github.com/vaamonde/ubuntu-1804/blob/master/owncloud.sh


CHANGELOG
Autor: Francisco Neto
Site: Francisco Neto
Facebook: facebook.com/# 
Facebook: facebook.com/# 
YouTube: youtube.com/# 
Data de criação: 29/01/2019
Data de atualização: 14/08/2020
Versão: 0.04
Testado e homologado para a versão do Ubuntu Server 18.04.x LTS x64
Kernel >= 4.15.x

============================================= ATUALIZAÇÕES DOS SCRIPTS =============================================
14/08/2020 - Francisco Neto - Criação dos scripts: fusion.sh e rocketchat.sh

03/08/2020 - Francisco Neto - Criação dos scripts: ansible.sh, glpi.sh, grafana.sh, graylog.sh, tomcat.sh, zabbix.sh.
03/08/2020 - Francisco Neto - Criação da Variável PTBR no script do loganalyzer.sh para adicionar o recurso de tradução
03/08/2020 - Francisco Neto - Criação do arquivo do Systemd do Portainer para inicializar automático o Container

24/07/2020 - Francisco Neto - Correção das novas dependências do Netdata, resolvido as falhas de dependência do Cloud.

22/07/2020 - Francisco Neto - Revisão de todos os scripts do repositório, atualização das versões dos aplicativos e 
correções de português nos scripts.

30/03/2020 - Francisco Neto - Atualização do arquivo de instalação do netdata.sh, acréscimo da dependência do libuv1-dev 

09/11/2019 - Francisco Neto - Atualização do arquivo de configuração sshd_config, acréscimo das linha MaxSessions 3, 
MaxStartups 5:60:10, AddressFamily inet, PrintLastLog yes

15/10/2019 - Robosn Vaamonde - Atualização do script webmin.sh, acréscimo da dependência unzip e correção da linha de versão do 
WEBMINVERSION=`echo webmin_*`;

28/02/2019 - Francisco Neto - Atualizado arquivo wp-config.php, acrescentado linhas para trabalhar com dois domínios ou dois
endereços IP diferentes no mesmo Wordpress;

25/02/2019 - Francisco Neto - Falha no script owncloud.sh, erro nas linhas de criação do banco de dados e na opção de 
descompactar o arquivo do owncloud.tar.bz2;

18/02/2019 - Francisco Neto - Informado falha no script webmin.sh, erro relacionado a linha 110, tinha uma barra de espaço
a mais no comando de apt -y install;

18/02/2019 - Francisco Neto - Informado falha no script netdata.sh, erro relacionado a linha 128, tinha uma barra de espaço
a mais no comando de apt -y install, acréscimo do bloco de verificação de dependências do MySQL e Apache2;

10/02/2019 - Francisco Neto - Atualização wordpress.sh, loganalyzer.sh, openfire.sh, zoneminder.sh, acréscimo do bloco de
verificação de dependências;

10/02/2019 - Francisco Neto - Atualização de todos os scripts, melhorias na parte de documentação dos scripts;

10/02/2019 - Francisco Neto - Atualização de todos os scripts, melhorias na parte de variável de arquivo de LOG;

29/01/2019 - Francisco Neto - Atualização de todos os scripts, melhorias na parte de validação do bloco IF;

29/01/2019 - Francisco Neto - Atualização de todos os scripts, melhorias na parte de variável de cálculo de tempo.


bareos.sh
#!/bin/bash
# Autor: Francisco Neto
# Site: www.softwarefn.com.br
# /# 
# /# 
# /# 
# Data de criação: 10/02/2019
# Data de atualização: 03/08/2020
# Versão: 0.04
# Testado e homologado para a versão do Ubuntu Server 18.04.x LTS x64
# Kernel >= 4.15.x
# Testado e homologado para a versão do BareOS 19.2.x
#
# O BareOS Server é uma bifurcação (fork) do projeto de código aberto Bacula versão 5.2. Em 2010, o desenvolvedor da 
# comunidade do Bacula, Marco Van Wieringen, começou a coletar contribuições das comunidades que foram rejeitadas ou 
# negligenciadas em começou a desenvolver sua própria versão. Este trabalho foi mais tarde a base da criação do software 
# BareOS e desde então foi enriquecido com vários novos recursos.
#
# O BareOS é um conjunto de programas de computador que permite ao administrador do sistema gerenciar o backup, a recuperação
# e a verificação de dados de computadores em uma rede de que tenha diferentes tipos de sistemas operacionais. O BareOS também
# pode ser executado inteiramente em um único computador e pode fazer backup para vários tipos de mídia, incluindo fita e 
# disco. O Bareos é relativamente fácil de usar e eficiente, além de oferecer muitos recursos avançados de gerenciamento de 
# armazenamento que facilitam a localização e a recuperação de arquivos perdidos ou danificados. Devido ao seu design modular,
# o BareOS é escalonável para pequenos sistemas de computadores individuais a sistemas que consistem em centenas de 
# computadores localizados em uma grande rede.
#
# Informações que serão solicitadas na configuração via Web do Bareos
# Please choose a director: localhost-dir
# Username: admin
# Password: bareos
# Please choose a language: Portuguese (BR): Login
#
# Soluções de Backup Open Source:
# Bacula: https://www.bacula.org/
# Amanda: http://www.amanda.org/
# Backup PC: http://backuppc.sourceforge.net/
#
# Site Oficial do Projeto: https://www.bareos.org/en/
#
# Vídeo de instalação do GNU/Linux Ubuntu Server 18.04.x LTS: https://www.youtube.com/watch?v=zDdCrqNhIXI
# Vídeo de instalação do LAMP Server no Ubuntu Server 18.04.x LTS: https://www.youtube.com/watch?v=6EFUu-I3
#
# Variável da Data Inicial para calcular o tempo de execução do script (VARIÁVEL MELHORADA)
# opção do comando date: +%T (Time)
HORAINICIAL=$(date +%T)
#
# Variáveis para validar o ambiente, verificando se o usuário é "root", versão do ubuntu e kernel
# opções do comando id: -u (user)
# opções do comando: lsb_release: -r (release), -s (short), 
# opões do comando uname: -r (kernel release)
# opções do comando cut: -d (delimiter), -f (fields)
# opção do shell script: piper | = Conecta a saída padrão com a entrada padrão de outro comando
# opção do shell script: acento crase ` ` = Executa comandos numa subshell, retornando o resultado
# opção do shell script: aspas simples ' ' = Protege uma string completamente (nenhum caractere é especial)
# opção do shell script: aspas duplas " " = Protege uma string, mas reconhece $, \ e ` como especiais
USUARIO=$(id -u)
UBUNTU=$(lsb_release -rs)
KERNEL=$(uname -r | cut -d'.' -f1,2)
#
# Variável do caminho do Log dos Script utilizado nesse curso (VARIÁVEL MELHORADA)
# opções do comando cut: -d (delimiter), -f (fields)
# $0 (variável de ambiente do nome do comando)
LOG="/var/log/$(echo $0 | cut -d'/' -f2)"
#
# Declarando a variável de download do BareOS (Link atualizado no dia 22/07/2020)
RELEASE="http://download.bareos.org/bareos/release/latest/xUbuntu_18.04/"
USER="admin"
PASSWD="bareos"
PROFILE="webui-admin"
POSTFIX="No configuration"
#
# Exportando o recurso de Noninteractive do Debconf para não solicitar telas de configuração
export DEBIAN_FRONTEND="noninteractive"
#
# Verificando se o usuário é Root, Distribuição é >=18.04 e o Kernel é >=4.15 <IF MELHORADO)
# [ ] = teste de expressão, && = operador lógico AND, == comparação de string, exit 1 = A maioria dos erros comuns na execução
clear
if [ "$USUARIO" == "0" ] && [ "$UBUNTU" == "18.04" ] && [ "$KERNEL" == "4.15" ]
	then
		echo -e "O usuário é Root, continuando com o script..."
		echo -e "Distribuição é >=18.04.x, continuando com o script..."
		echo -e "Kernel é >= 4.15, continuando com o script..."
		sleep 5
	else
		echo -e "Usuário não é Root ($USUARIO) ou Distribuição não é >=18.04.x ($UBUNTU) ou Kernel não é >=4.15 ($KERNEL)"
		echo -e "Caso você não tenha executado o script com o comando: sudo -i"
		echo -e "Execute novamente o script para verificar o ambiente."
		exit 1
fi
#
# Verificando se as dependências do BareOS estão instaladas
# opção do dpkg: -s (status), opção do echo: -e (interpretador de escapes de barra invertida), -n (permite nova linha)
# || (operador lógico OU), 2> (redirecionar de saída de erro STDERR), && = operador lógico AND, { } = agrupa comandos em blocos
# [ ] = testa uma expressão, retornando 0 ou 1, -ne = é diferente (NotEqual)
echo -n "Verificando as dependências, aguarde... "
	for name in mysql-server mysql-common apache2
	do
  		[[ $(dpkg -s $name 2> /dev/null) ]] || { echo -en "\n\nO software: $name precisa ser instalado. \nUse o comando 'apt install $name'\n";deps=1; }
	done
		[[ $deps -ne 1 ]] && echo "Dependências.: OK" || { echo -en "\nInstale as dependências acima e execute novamente este script\n";exit 1; }
		sleep 5
#
# Exportando o recurso de Noninteractive do Debconf para não solicitar telas de configuração
export DEBIAN_FRONTEND="noninteractive"
#
# Script de instalação do BareOS no GNU/Linux Ubuntu Server 18.04.x
# opção do comando echo: -e (enable interpretation of backslash escapes), \n (new line)
# opção do comando hostname: -I (all IP address)
# opção do comando date: + (format), %d (day), %m (month), %Y (year 1970), %H (hour 24), %M (minute 60)
# opção do comando cut: -d (delimiter), -f (fields)
echo -e "Início do script $0 em: `date +%d/%m/%Y-"("%H:%M")"`\n" &>> $LOG
clear
#
echo
echo -e "Instalação do BareOS no GNU/Linux Ubuntu Server 18.04.x\n"
echo -e "Após a instalação do BareOS acessar a URL: http://`hostname -I | cut -d' ' -f1`/bareos-webui\n"
echo -e "Aguarde, esse processo demora um pouco dependendo do seu Link de Internet...\n"
sleep 5
#
echo -e "Adicionando o Repositório Universal do Apt, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	add-apt-repository universe &>> $LOG
echo -e "Repositório adicionado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Adicionando o Repositório Multiversão do Apt, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	add-apt-repository multiverse &>> $LOG
echo -e "Repositório adicionado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Atualizando as listas do Apt, aguarde..."
	#opção do comando: &>> (redirecionar a saída padrão)
	apt update &>> $LOG
echo -e "Listas atualizadas com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Atualizando o sistema, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando apt: -y (yes)
	apt -y upgrade &>> $LOG
echo -e "Sistema atualizado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Removendo software desnecessários, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando apt: -y (yes)
	apt -y autoremove &>> $LOG
echo -e "Software removidos com sucesso!!!, continuando com o script..."
sleep 5
#
echo -e "Instalando o BareOS, aguarde...\n"
#
echo -e "Adicionando o repositório do BareOS, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando cp: -v (verbose)
	# opção do comando wget: -q -O- (file)
	# opção do redirecionador |: Conecta a saída padrão com a entrada padrão de outro comando
	# opção do comando apt-key: add (file name), - (arquivo recebido dO redirecionar | piper)
	cp -v conf/bareos.list /etc/apt/sources.list.d/bareos.list &>> $LOG
	wget -q $RELEASE/Release.key -O- | apt-key add - &>> $LOG
	apt update &>> $LOG
echo -e "Repositório do BareOS criado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Configurando as variáveis do Debconf do BareOS para o Apt, aguarde..."
	# opção do comando: &>> (redirecionar a saída padrão)
	# opção do comando | (piper): (Conecta a saída padrão com a entrada padrão de outro comando)
	echo "bareos-database-common bareos-database-common/dbconfig-install boolean true" | debconf-set-selections
	echo "bareos-database-common bareos-database-common/mysql/app-pass password $PASSWD" | debconf-set-selections
	echo "bareos-database-common bareos-database-common/app-password-confirm password $PASSWD" | debconf-set-selections
	echo "postfix postfix/main_mailer_type string $POSTFIX" | debconf-set-selections
	debconf-show bareos-database-common &>> $LOG
	debconf-show postfix &>> $LOG
echo -e "Variáveis configuradas com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Instalando o BareOS Server e criando a Base de Dados em MySQL, aguarde..."
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando apt: -y (yes)
	apt -y install bareos bareos-database-mysql bareos-tools bareos-bconsole &>> $LOG
echo -e "BareOS instalado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Instalando o BareOS Webgui, aguarde..."
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando apt: -y (yes)
	apt -y install bareos-webui &>> $LOG
echo -e "BareOS Webgui instalado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Reinicializando os Serviços do BareOS Server, aguarde..."
	# bareos-dir: Bareos Director.
	# bareos-sd: Bareos Storage Daemon.
	# bareos-fd: Bareos File Daemon
	# bareos-webui: Bareos Web Configurator
	systemctl start bareos-dir.service &>> $LOG
	systemctl start bareos-sd.service &>> $LOG
	systemctl start bareos-fd.service &>> $LOG
	systemctl restart apache2.service &>> $LOG
echo -e "Serviços reinicializados com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Criando o usuário de administração do BareOS Webgui, aguarde..."
	# opção do comando: &>> (redirecionar a saida padrão)
	# opção do comando | (piper): (Conecta a saída padrão com a entrada padrão de outro comando)
	# opções do comando bconsole: configure (configurar o BareOS), add (adicionar), console (gerenciamento do console)
	# name (nome do usuário), password (senha do usuário), profile (perfil do usuário), tlsenable (habilitar ou desabilitar TLS)
	# reload (aplicar as mudanças)
	echo -e "configure add console name=$USER password=$PASSWD profile=$PROFILE tlsenable=no" | bconsole &>> $LOG
	echo -e "reload" | bconsole &>> $LOG
echo -e "Usuário criado com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Verificando as portas de Conexões do BareOS, aguarde..."
	# opção do comando netstat: -a (all), -n (numeric)
	netstat -an | grep '9102\|9103'
echo -e "Portas de conexões verificadas com sucesso!!!, continuando com o script..."
sleep 5
echo
#
echo -e "Instalação do BareOS feita com Sucesso!!!."
	# script para calcular o tempo gasto (SCRIPT MELHORADO, CORRIGIDO FALHA DE HORA:MINUTO:SEGUNDOS)
	# opção do comando date: +%T (Time)
	HORAFINAL=$(date +%T)
	# opção do comando date: -u (utc), -d (date), +%s (second since 1970)
	HORAINICIAL01=$(date -u -d "$HORAINICIAL" +"%s")
	HORAFINAL01=$(date -u -d "$HORAFINAL" +"%s")
	# opção do comando date: -u (utc), -d (date), 0 (string command), sec (force second), +%H (hour), %M (minute), %S (second), 
	TEMPO=$(date -u -d "0 $HORAFINAL01 sec - $HORAINICIAL01 sec" +"%H:%M:%S")
	# $0 (variável de ambiente do nome do comando)
	echo -e "Tempo gasto para execução do script $0: $TEMPO"
echo -e "Pressione <Enter> para concluir o processo."
# opção do comando date: + (format), %d (day), %m (month), %Y (year 1970), %H (hour 24), %M (minute 60)
echo -e "Fim do script $0 em: `date +%d/%m/%Y-"("%H:%M")"`\n" &>> $LOG
read
exit 1
